cpp/include/cuvs/neighbors/tiered_index.hpp

```diff
@@ -197,4 +197,30 @@ void search(raft::resources const& res,
             raft::device_matrix_view<float, int64_t, raft::row_major> distances,
             const cuvs::neighbors::filtering::base_filter& sample_filter =
               cuvs::neighbors::filtering::none_sample_filter{});
+
+/** @brief Merge multiple tiered indices into a single index.
+ *
+ * This function merges multiple tiered indices into one, combining both the datasets and graph
+ * structures.
+ *
+ * @param[in] res
+ * @param[in] indices A vector of pointers to the indices to merge. All indices should
+ *                    be of the same type, and have datasets with the same dimensionality
+ *
+ * @return A new tiered index containing the merged indices
+ */
+auto merge(raft::resources const& res,
```

Comment:
We discussed this briefly offline, but I agree it can be kind of awkward to reason about what a merged tiered index would do, and even more awkward- what it would look like. I think instead we should consider the user having to merge the underlying indexes at each of the tiers and provide new fsctory functions that can take those new merged indexes (whether physical or logical) and construct a new tiered index with them. Ideally, the bfknn would only do a physical merge, but I suppose we could always support  a logical merge and if more points are added, we create a new bfknn index in the composite and just add to that?

Reply:
Update: After discussion this more offiline, we've decided to remove the merge strategies altogether and only support physical merges in the "merge()" apis. The concept of logical merge might still be done in future APIs (likely through the use of something like a composite_index), but that won't be part of the official merge() API.